### ISSUE TITLE
Problem on AWICM3 installation on JUWELS

### ISSUE THREAD

**Issue by shijian0702** [esm-tools/esm_tools#593]
Hi all,

I have some problems when installing AWICM3 on JUWELS, which I think may be related to the netcdf-Fortran configuration.
I will be grateful if anyone can have a look at it. By the way, this issue cannot be reproduced in other HPCs, like ollie, mistral.

Best,
Jian

**System**
 - Supercomputer: JUWELS (Stages 2022, Python/3.9.6, netCDF-Fortran/4.5.3-serial)
 - ESM-Tools 6.1**
 
I used the command: **esm_master install-awicm3-deck**, and it shows errors after asking for permission of all the components: 
(I also alter different stages, python, netcdf versions, but the same errors remain)

-- setting default build type: Release
-- The C compiler identification is Intel 2021.4.0.20210910
-- The CXX compiler identification is Intel 2021.4.0.20210910
-- Detecting C compiler ABI info
-- Detecting C compiler ABI info - done
-- Check for working C compiler: /p/software/juwels/stages/2022/software/OpenMPI/4.1.2-intel-compilers-2021.4.0/bin/mpicc - skipped
-- Detecting C compile features
-- Detecting C compile features - done
-- Detecting CXX compiler ABI info
-- Detecting CXX compiler ABI info - done
-- Check for working CXX compiler: /p/software/juwels/stages/2022/software/OpenMPI/4.1.2-intel-compilers-2021.4.0/bin/mpic++ - skipped
-- Detecting CXX compile features
-- Detecting CXX compile features - done
-- The Fortran compiler identification is Intel 2021.4.0.20210910
-- Detecting Fortran compiler ABI info
-- Detecting Fortran compiler ABI info - done
-- Check for working Fortran compiler: /p/software/juwels/stages/2022/software/OpenMPI/4.1.2-intel-compilers-2021.4.0/bin/mpifort - skipped
-- Checking whether /p/software/juwels/stages/2022/software/OpenMPI/4.1.2-intel-compilers-2021.4.0/bin/mpifort supports Fortran 90
-- Checking whether /p/software/juwels/stages/2022/software/OpenMPI/4.1.2-intel-compilers-2021.4.0/bin/mpifort supports Fortran 90 - yes
-- Looking for Fortran NCCRE
-- Looking for Fortran NCCRE - not found
-- Looking for nc__create
-- Looking for nc__create - not found
-- Configuring done
**CMake Error: The following variables are used in this project, but they are set to NOTFOUND.**
Please set them or make sure they are set and tested correctly in the CMake files:
/p/home/jusers/shi4/juwels/model_codes/awicm3-deck/oasis/lib/psmile/NETCDF_Fortran_INCLUDE_DIRECTORIES
   used as include directory in directory /p/home/jusers/shi4/juwels/model_codes/awicm3-deck/oasis/lib/psmile
/p/home/jusers/shi4/juwels/model_codes/awicm3-deck/oasis/lib/scrip/src/NETCDF_Fortran_INCLUDE_DIRECTORIES
   used as include directory in directory /p/home/jusers/shi4/juwels/model_codes/awicm3-deck/oasis/lib/scrip/src
NETCDF_Fortran_LIBRARIES
    linked by target "psmile" in directory /p/home/jusers/shi4/juwels/model_codes/awicm3-deck/oasis/lib/psmile

-- Generating done
CMake Generate step failed.  Build files cannot be regenerated correctly.
Traceback (most recent call last):
  File "/p/home/jusers/shi4/juwels/.local/bin/esm_master", line 33, in <module>
    sys.exit(load_entry_point('esm-tools', 'console_scripts', 'esm_master')())
  File "/p/home/jusers/shi4/juwels/esm_tools/src/esm_master/cli.py", line 103, in main
    main_flow(parsed_args, target)
  File "/p/home/jusers/shi4/juwels/esm_tools/src/esm_master/esm_master.py", line 106, in main_flow
    user_task.execute(ignore_errors)  # env)
  File "/p/home/jusers/shi4/juwels/esm_tools/src/esm_master/task.py", line 494, in execute
    subprocess.run(
  File "/p/software/juwels/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['./comp-oasis3mct-4.0-oifsdeck_script.sh']' returned non-zero exit status 1.

**Loaded modules**
![image](https://user-images.githubusercontent.com/18346694/156032415-50a63556-4f38-4c69-83c7-7a239530612c.png)

**Comment by mandresm** [comment]
Hi @shijian0702 ,

I thought that the `deck` version was going to be discontinued. @JanStreffing is that right? In that case what version should @shijian0702 be using?

**Comment by JanStreffing** [comment]
Yes, the DECK version is indeed not the version used for the DECK paper. Instead it is what we started with a year and a half ago. It is out of date and contains many errors we fixed for v3.0. I suggest we delete it asap.

@shijian0702 please try to install either awicm3-v3.0 for the version I use in my DECK paper, or awicm3-frontiers-xios for the version that @tsemmler05 is currently tuning, and which results in the best simulations so far.

_Interestingly the error you stumbled over has nothing to do with the awicm3 version, but rather that it tired to install awicm3 with libraries from a GEOMAR project folder that you had no access to. You might get the same issue with the other model version. In that case we will look closer._
edit: I misread that.

**Comment by mandresm** [comment]
When removing `deck`, can I also remove the following versions from their respective yamls?
  - oifs-43r3-awicm-deck      
  - fesom-2.0-jio                                                                                       
  - oasis3mct-4.0-oifsdeck

**Comment by JanStreffing** [comment]
I checked, they don't seem to be used anywhere else. I think they are good for removal as well.

**Comment by ackerlar** [comment]
Hi all,
I get the same error when trying to install awiesm-2.1 on JUWELS

```
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/psmile/NETCDF_Fortran_INCLUDE_DIRECTORIES
   used as include directory in directory /p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/psmile
/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/scrip/src/NETCDF_Fortran_INCLUDE_DIRECTORIES
   used as include directory in directory /p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/scrip/src
NETCDF_Fortran_LIBRARIES
    linked by target "psmile" in directory /p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/psmile

-- Configuring incomplete, errors occurred!
See also "/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/build/CMakeFiles/CMakeOutput.log".
See also "/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/build/CMakeFiles/CMakeError.log".
Traceback (most recent call last):
  File "/p/home/jusers/ackermann1/juwels/.local/bin/esm_master", line 33, in <module>
    sys.exit(load_entry_point('esm-tools', 'console_scripts', 'esm_master')())
  File "/p/home/jusers/ackermann1/juwels/esm_tools/src/esm_master/cli.py", line 103, in main
    main_flow(parsed_args, target)
  File "/p/home/jusers/ackermann1/juwels/esm_tools/src/esm_master/esm_master.py", line 106, in main_flow
    user_task.execute(ignore_errors)  # env)
  File "/p/home/jusers/ackermann1/juwels/esm_tools/src/esm_master/task.py", line 494, in execute
    subprocess.run(
  File "/p/software/juwels/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['./comp-oasis3mct-2.8-paleodyn_script.sh']' returned non-zero exit status 1.
```

 - loaded modules: 
 ```
Currently Loaded Modules:
  1) Stages/2022     (S)   5) StdEnv/2022            9) Tcl/8.6.11        13) libffi/.3.4.2   (H)  17) libyaml/.0.2.5         (H)  21) NASM/.2.15.05        (H)
  2) GCCcore/.11.2.0 (H)   6) bzip2/.1.0.8     (H)  10) SQLite/.3.36 (H)  14) OpenSSL/1.1          18) PostgreSQL/13.4             22) libjpeg-turbo/.2.1.1 (H)
  3) zlib/.1.2.11    (H)   7) ncurses/.6.2     (H)  11) XZ/.5.2.5    (H)  15) libxml2/.2.9.10 (H)  19) gflags/.2.2.2          (H)  23) Python/3.9.6
  4) binutils/.2.37  (H)   8) libreadline/.8.1 (H)  12) GMP/6.2.1         16) libxslt/.1.1.34 (H)  20) libspatialindex/.1.9.3 (H)

  Where:
   S:  Module is Sticky, requires --force to unload or purge
   H:             Hidden Module
```
 - esm_tools version 6.1

**Comment by mandresm** [comment]
@ackerlar , I'm afraid AWI-ESM is not supported in Juwels. If you'd like to have support there I suggest you open an independent issue for that, but I think as long as we have to solve all the versioning issues of AWIESM currently going on, we cannot handle the implementation of AWIESM in Juwels due to lack of person time. If you want to give it a try please go ahead, but I think it's gonna be difficult for us even to just provide help.

**Comment by JanStreffing** [comment]
I just chatted with Lars. The problem from him was that in `juwels.yaml` we set a default:
`compiler_mpi: intel2019_impi2019`
For awi-esm-2.1 this is never overwritten and we never define the NetCDF env vars for `intel2019_impi2019` `awi_libs` (only for `geomar_libs`)
For awi-cm3 this is not a problem cause we overwrite with
`compiler_mpi: intel2022_ompi2022` in the `awicm3.yaml`. And for that we have defined the paths. The reason @shijian0702 could still not install awi-cm3, is that the paths point towards the Climate dynamics section project chhb19. I assume Jian is part of the Paleo project chhb20. This is a problem because of the restrictive settings on juwels that don't even give read access to folders that are not in your project.

I will install the libs there again for chhb20. For us this means it's time to split `awi_libs` int `chhb19_libs` and `chhb20_libs`.

**Comment by mandresm** [comment]
Thanks @JanStreffing , I don't think is going to be necessary to "split" the libs. In the end it's only one section of the path that changes there right? Maybe we can handle this easily with a variable there, for example:
```
project: c${general.account}
```
and then for all the paths containing `chhb19` changing them to:
```
IO_LIB_ROOT: /p/project/${project}/streffing1/software/HPC_libraries/intel-2021.4.0_openmpi-4.1.2_20220129
```

**Comment by JanStreffing** [comment]
Jup that works. I should probably remove my name from that path. It's not required. I could install in e.g.

`/p/project/${project}/HPC_libraries/intel-2021.4.0_openmpi-4.1.2_20220129`

Probably best done with links for now. This way others, like Jan Hegewald can also install libs.

**Comment by mandresm** [comment]
Should we do this in the branch of the open PR #594?

**Comment by JanStreffing** [comment]
It's a different issue, but sure, we can. I'm installing the libs on chhb20 now, and will make the links. Once they are both available I will post them.

**Comment by seb-wahl** [comment]
I would avoid moving or linking the libraries from `/p/project/${project}/streffing1/software/HPC_libraries/intel-2021.4.0_openmpi-4.1.2_20220129` to `/p/project/${project}/HPC_libraries/intel-2021.4.0_openmpi-4.1.2_20220129` but instead completely re-install in the new path `/p/project/${project}/HPC_libraries/intel-2021.4.0_openmpi-4.1.2_20220129` I have had issues in the past when moving libs around as I think some dependencies use the full path. Just to be on the safe side.

**Comment by JanStreffing** [comment]
Good advice. I will install a new set for chhb19. I already installed the ones in chhb20 right in the new location.

**Comment by shijian0702** [comment]
yes, @JanStreffing, @mandresm, @seb-wahl,  I've tried other awicm3 version, and the problems remains. I‘m indeed in chhb20. Thanks for your efforts:)

**Comment by mandresm** [comment]
> Good advice. I will install a new set for chhb19. I already installed the ones in chhb20 right in the new location.

@JanStreffing let me know when I can proceed with the `${project}` modifications. All the tests passed for #594 in Ollie and Aleph. I'll do the same in Juwels once I have the new commits.

**Comment by seb-wahl** [comment]
To avoid conflicts in `juwels.yaml` later on, you may want to merge #595 into your dev branch first. #595 adds new sections for Stage/2022 with ParaStationMPI

**Comment by ackerlar** [comment]
I still get the same error:
```
-- Check for working Fortran compiler: /p/software/juwels/stages/2022/software/OpenMPI/4.1.2-intel-compilers-2021.4.0/bin/mpifort - skipped
-- Checking whether /p/software/juwels/stages/2022/software/OpenMPI/4.1.2-intel-compilers-2021.4.0/bin/mpifort supports Fortran 90
-- Checking whether /p/software/juwels/stages/2022/software/OpenMPI/4.1.2-intel-compilers-2021.4.0/bin/mpifort supports Fortran 90 - yes
-- Looking for Fortran NCCRE
-- Looking for Fortran NCCRE - not found
-- Looking for nc__create
-- Looking for nc__create - not found
-- Configuring done
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/psmile/NETCDF_Fortran_INCLUDE_DIRECTORIES
   used as include directory in directory /p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/psmile
/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/scrip/src/NETCDF_Fortran_INCLUDE_DIRECTORIES
   used as include directory in directory /p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/scrip/src
NETCDF_Fortran_LIBRARIES
    linked by target "psmile" in directory /p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/psmile
```

I changed the default in `juwels.yaml` to `compiler_mpi: intel2022_ompi2022` and this to:
```
    awi_libs:
       choose_compiler_mpi:
          intel2022_ompi2022:
             add_export_vars:
                #IO_LIB_ROOT: /p/project/chhb19/streffing1/software/HPC_libraries/intel-2021.4.0_openmpi-4.1.2_20220129
                IO_LIB_ROOT: /p/project/chhb20/HPC_libraries/intel-2021.4.0_openmpi-4.1.2_20220129
```

**Comment by JanStreffing** [comment]
The folders should be ready by now. @mandresm

**Comment by mandresm** [comment]
@ackerlar what esm_master command are you using to compile?

**Comment by ackerlar** [comment]
I use `esm_master comp-awiesm-2.1` but I also tried from scratch with `esm_master install-awiesm-2.1` after removing the old awiesm-2.1 directory

**Comment by JanStreffing** [comment]
@shijian0702 I installed the libraries in a path that is accessible to you (`/p/project/chhb19/HPC_libraries/intel-2021.4.0_openmpi-4.1.2_20220129`)
@mandresm Can we potentially extract the users group information via:
```
echo "groups:" ; for i in $(id -Gn $whoami);do echo "  - $i" ;done
groups:
  - jusers
  - hhb19
  - juwels
  - juwels_gpus
  - cesmtst
  - chhb19
  - chhb20
  - slmet
  - chhb21
  - esmpart
```
All AWI projects are called chhb?? so far.

If not we maybe have to rely on the user setting a variable $PROJECT for us in this manner in the .bash_rc.
: https://esm-tools.readthedocs.io/en/latest/ten_steps.html 

Then we can use this in the juwels.yaml
I will make a branch that reads the variable onces it is set.

**Comment by mandresm** [comment]
I think the user has to specify it with an environment variable because a user can be in more than one project (that's my case)

**Comment by JanStreffing** [comment]
Through some bash magic we can extract the first group name that starts with chhb. It's not fool-proof, but it will cover 95% of cases. And it allows us not to make the readthedocs any more complicated that in should be.

@shijian0702 & @ackerlar: You can test out https://github.com/esm-tools/esm_tools/tree/fix/awi_libs_and_project_juwels, or wait for the merge, as you wish. No guarantee for AWI-ESM2.1 obviously.

**Comment by shijian0702** [comment]
@JanStreffing 

I totally clean the old eam_tools and install the latest version, then the esm_master says:

**[shi4@jwlogin02 model_codes]$ esm_master** 

**No targets found for keyword nemo-ORCA05_LIM2_KCM_AOW_FS+echam-6.3.05p2-foci. Type 'esm_master' to get a full list
of available targets.**


Though try again, esm_master won't get a full list of available models.

**Comment by JanStreffing** [comment]
That's one for the esm_tools devs. Probably something went wrong during the uninstall & reinstall.

**Comment by ackerlar** [comment]
I could overcome this by just commenting out the corresponding lines in `esm_tools/configs/setups/foci/foci.yaml`
```
        choose_version:
          default:
            couplings:
            - nemo-ORCA05_LIM2_KCM_AOW+echam-6.3.05p2-foci
              #fs:
              #  couplings:
              #  - nemo-ORCA05_LIM2_KCM_AOW_FS+echam-6.3.05p2-foci
          default_oasismct4:
            couplings:
```

Of course not nice but for testing okay, I guess

**Comment by JanStreffing** [comment]
May be caused by this recent merge? https://github.com/esm-tools/esm_tools/pull/599 @seb-wahl @pgierz

**Comment by ackerlar** [comment]
Is `PROJECT_ID` supposed to be set automatically? 

In any case, I run:
```
export PROJECT_ID=chhb20
esm_master clean-awiesm-2.1
esm_master comp-awiesm-2.1
```

and get the same error as before:

```
-- Looking for Fortran NCCRE
-- Looking for Fortran NCCRE - not found
-- Looking for nc__create
-- Looking for nc__create - not found
CMake Error: The following variables are used in this project, but they are set to NOTFOUND.
Please set them or make sure they are set and tested correctly in the CMake files:
/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/psmile/NETCDF_Fortran_INCLUDE_DIRECTORIES
   used as include directory in directory /p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/psmile
/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/scrip/src/NETCDF_Fortran_INCLUDE_DIRECTORIES
   used as include directory in directory /p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/scrip/src
NETCDF_Fortran_LIBRARIES
    linked by target "psmile" in directory /p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/lib/psmile

-- Configuring incomplete, errors occurred!
See also "/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/build/CMakeFiles/CMakeOutput.log".
See also "/p/home/jusers/ackermann1/juwels/awiesm-2.1/oasis/build/CMakeFiles/CMakeError.log".
```

**Comment by JanStreffing** [comment]
Only if you follow the updated user guide: https://github.com/esm-tools/esm_tools/commit/5407c3733162379d3ebfd91fc57d7d0dfe53b7ee
Just to make sure it's not an awi-esm-2.1 related issue, can you install awicm3-frontiers?

**Comment by ackerlar** [comment]
No, but I guess it is because I don't have access to the repositories?

```
[ackermann1@jwlogin07 ~]$ esm_master comp-awicm3-frontiers/oasis
    Executing commands in this order:
         mkdir -p awicm3-frontiers
         cd awicm3-frontiers
         sed -i '/COUPLENEMOECE = /s/.TRUE./.FALSE./g' oifs-43r3/src/ifs/module/yommcc.F90
         sed -i '/COUPLEFESOM2 = /s/.FALSE./.TRUE./g' oifs-43r3/src/ifs/module/yommcc.F90
         sed -i '/COUPLENEMOFOCI = /s/.TRUE./.FALSE./g' oifs-43r3/src/ifs/module/yommcc.F90
         cd oasis
         mkdir -p build; cd build; cmake ..;   make -j 1; mkdir -p ../include/; cp lib/psmile/libpsmile.a lib/psmile/mct/libmct.a lib/psmile/mct/mpeu/libmpeu.a lib/psmile/scrip/libscrip.a ../lib; cp lib/psmile/mod_oasis*mod ../include
         cd ..
         mkdir -p ./lib
         cp  oasis/build/lib/psmile/libpsmile.a  lib
         cp  oasis/build/lib/psmile/mct/libmct.a  lib
         cp  oasis/build/lib/psmile/mct/mpeu/libmpeu.a  lib
         cp  oasis/build/lib/psmile/scrip/libscrip.a  lib
         cd ..
sed: can't read oifs-43r3/src/ifs/module/yommcc.F90: No such file or directory
Traceback (most recent call last):
  File "/p/home/jusers/ackermann1/juwels/.local/bin/esm_master", line 33, in <module>
    sys.exit(load_entry_point('esm-tools', 'console_scripts', 'esm_master')())
  File "/p/home/jusers/ackermann1/juwels/esm_tools/src/esm_master/cli.py", line 103, in main
    main_flow(parsed_args, target)
  File "/p/home/jusers/ackermann1/juwels/esm_tools/src/esm_master/esm_master.py", line 106, in main_flow
    user_task.execute(ignore_errors)  # env)
  File "/p/home/jusers/ackermann1/juwels/esm_tools/src/esm_master/task.py", line 494, in execute
    subprocess.run(
  File "/p/software/juwels/stages/2022/software/Python/3.9.6-GCCcore-11.2.0/lib/python3.9/subprocess.py", line 528, in run
    raise CalledProcessError(retcode, process.args,
subprocess.CalledProcessError: Command '['sed', '-i', '/COUPLENEMOECE = /s/.TRUE./.FALSE./g', 'oifs-43r3/src/ifs/module/yommcc.F90']' returned non-zero exit status 2.
```

**Comment by JanStreffing** [comment]
I've added you.

**Comment by shijian0702** [comment]
After commenting out these lines @ackerlar suggested, I succeed to install awicm3-v3.0 with esm-tools/fix/awi_libs_and_project_juwels

Thanks a lot guys!

**Comment by ackerlar** [comment]
Thanks @JanStreffing ! At least, it compiled without errors. I see how far I come with awiesm

**Comment by seb-wahl** [comment]
> @JanStreffing
> 
> I totally clean the old eam_tools and install the latest version, then the esm_master says:
> 
> **[shi4@jwlogin02 model_codes]$ esm_master**
> 
> **No targets found for keyword nemo-ORCA05_LIM2_KCM_AOW_FS+echam-6.3.05p2-foci. Type 'esm_master' to get a full list of available targets.**
> 
> Though try again, esm_master won't get a full list of available models.

This has been fixed in #601 that will be merged today. During the merge of "my" branch into release git decided not to add certain file. Sorry for the trouble.

**Comment by JanStreffing** [comment]
Thanks for taking care of it quickly!
